commons/src/main/java/org/infinispan/commons/api/functional/Listeners.java

```diff
@@ -0,0 +1,221 @@
+package org.infinispan.commons.api.functional;
+
+import org.infinispan.commons.api.functional.EntryView.ReadEntryView;
+
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Listener definitions.
+ *
+ * DESIGN RATIONALE:
+ * <ul>
+ *    <li>The current set of listener events that can be fired are related
+ *    to modifications, and deciding between create or modify, or providing
+ *    removed entry information, require reading the previous value. Hence,
+ *    the current set of listeners can only be provided for read-write maps.
+ *    In the future, if for example cache entry visited events are to be
+ *    supported, those would need to be associated with either a read or
+ *    read-write map, and hence the listeners interface would most likely
+ *    be split up.
```

Comment:
@wburns Currently the events are only. This is mostly due to the convenience from an implementation POV, because different commands implement `perform()` which happens before the entries are committed to container. Ideally, I think we'd want post only (JCache does this), but the problem is that it doesn't seem very obvious how each command can send post events, since these are fired from ClusteringDependentLogic implementations. Needs some thought.